### ISSUE TITLE
Set environmental variables?

### ISSUE THREAD

**Issue by ErichDonGubler** [cbourjau/cargo-with#22]
For a lot of Rust development that I do, it's particularly awesome to be able to set some environmental variables for test runs where extra logging and/or backtraces might be helpful:

```
cargo build && RUST_BACKTRACE=1 RUST_LOG=trace ./target/debug/thing --arg
```

My intuition tells me that something like this would be what I want:

```
cargo with "RUST_BACKTRACE=1 RUST_LOG=trace {bin} {args}" -- run
```

...but `RUST_BACKTRACE=1` isn't a binary, so right now `cargo-with` understandably panics:

```
thread 'main' panicked at 'Failed to spawn child process: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:999:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

It would be great if there were a facility for adding environmental variables!

**Comment by cbourjau** [comment]
Sorry, I'm not sure I understand your use case exactly. You **don't** want to run the built binaries through a third party program but rather run them as they are except with different environment variables set? If so, why not just run it with vanilla cargo? The following should work out of the box:
```
RUST_BACKTRACE=1 RUST_LOG=trace cargo run -- [your args here]
```
If you do want to run it through a third party program such as gdb and want to set the environment variables the following should just work:
```
RUST_BACKTRACE=1 cargo with "gdb --args" -- run --  [your args here]
```

In other words, the environment variables are just passed through. Or did I misunderstand you and the entire issue was more about the UI?

**Comment by ErichDonGubler** [comment]
You're understanding what I'm saying, methinks, so let me address the point about vanilla Cargo. :) The reason I don't want to use `RUST_LOG=something cargo run ...` specifically is because this will often add a bunch of logging I simply don't care about when specific `build.rs` files use the `env_logger` (which, surprisingly, isn't uncommon where I work -- odd, I know). The concern applies much less for other variables (e.g., `OAUTH_APP_SECRET=...`), but even so it'd still be nice to know that the environment is separated.

Does that make more sense? It IS a niche use case -- and I would understand if you viewed it as an expansion of scope you don't want in `cargo-with`.

**Comment by cbourjau** [comment]
Right now, the environment is not separated, though. The child process is spawned with all the variables of the parent (cargo-with) process. I think that behavior is the correct default. Allowing to set additional variables only for the child process seems to suggest that there is some sort of isolation in the first place, which is not the case.

Either way, concerning your use case, I guess you tried setting the module for the logger according to https://docs.rs/env_logger/0.6.2/env_logger/#enabling-logging ?

Other than that, did you consider something along the lines of (untested):
```
RUST_LOG=trace `cargo with "echo {bin}" -- run`  [your args here]
``` 
i.e., you just have cargo-with return you the the path to the binary and then execute it yourself

**Comment by ErichDonGubler** [comment]
Let me see if I can summarize in a different way: **I want a cross-platform  [`env`](https://en.wikipedia.org/wiki/Env)** to use with `cargo-with`. In a *nix environment, I can use `cargo-with` like this to get what I'm hoping for:

```
cargo with "env RUST_LOG=trace {bin} {args}" -- run -- --some-arg
```

However, no such convenience binary exists like this in Windows. It could be argued that that is outside of the scope of `cargo-with`, though. 

---

Responding to you above:

* Setting a module isn't exactly what I want either -- I DO want all `trace` logging output (and maybe set other variables), but only for the run of the specific binary I'm building. If I set it in the environment used by whatever ends up compiling that binary, then I get junk I don't want.
* The note about environments being separate was only meant in reference to new environmental variables I wanted to specify -- I do understand that the environment of child processes is expected to be shared with that of their parent. :)

**Comment by barskern** [comment]
For adding specific environment variables to the program run through `cargo-with`, we could simply add a command line argument;

```
cargo with --env "RUST_LOG=trace" "{bin} {args}" -- run
```

What is passed to `--env` could be passed to the command ([here](https://github.com/cbourjau/cargo-with/blob/55a8204d772521be923b7e8d96c2171726896a70/src/runner.rs#L67-L71)) through [`Command::env`](https://doc.rust-lang.org/std/process/struct.Command.html#method.env) or [`Command::envs`](https://doc.rust-lang.org/std/process/struct.Command.html#method.envs). This would not be hard to implement, but it is quite a niche usecase.

Would this satisfy your usecase @ErichDonGubler ?

**Comment by ErichDonGubler** [comment]
@barskern: That would! That sort of option would have been my first suggestion -- I just wanted to make sure the problem was understood before proposing it.

**Comment by barskern** [comment]
I have implemented this in #23. Perhaps you could give it a go to see if it works for you @ErichDonGubler ? :smile:

**Comment by cbourjau** [comment]
Sorry, busy weekend! I am a bit worried about feature creep here and would prefer it if we could get around it. Did you try:
```
RUST_LOG=trace `cargo with "echo {bin}" -- run`  [your args here]
```

**Comment by ErichDonGubler** [comment]
@cbourjau, your proposed solution still doesn't satisfy Windows usage, which is one of my use cases. Backtick script invocation is a POSIX shell feature, and unfortunately isn't available in the Windows world.

I've been thinking of writing a tool that emulates `env`, which could be used to solve this particular issue. It'd be used just like POSIX `env`:

```sh
# Using `envr` as a strawman name:
cargo with "envr RUST_LOG=trace {bin} {args}" -- run -- <args>
```

If a solution isn't applied here then I'm motivated enough to write such a utility. I do think, though, that some docs suggesting using `env` on *nix platforms in the meantime would be helpful for people running into this use case -- I doubt I'm the only one.

**Comment by cbourjau** [comment]
The more I think about it the more I agree with your initial statement that this is outside of the scope of this tool. If we take a step back, what we want to be doing is:
- Run `cargo with echo -- run <...>` to get the path to the binary with the right arguments 
- Set the environment variables and execute the the binary from the string

Sure, we could integrate this into `cargo-with` as shown by @barskern but I think that this use-case can be served better with some shell-foo.  Unfortunately, I have no scripting experience on Windows so I don't really know how a working example would look like. Including a section in the readme explaining the environment of the child process and a possible workaround for your usecase is certainly a good idea. If you were to figure out a working solution I am very interested to see it! Regardless of all this, I think that a proper cross-platform `env` would be a really great project by itself!

**Comment by cbourjau** [comment]
I'll close this for now. Would be nice to see a shell or `envr` workaround to add to the readme.
